// @azure/ai-anomaly-detector/dist-esm/src/models.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export var KnownTimeGranularity;
(function (KnownTimeGranularity) {
    /**
     * Yearly time granularity
     */
    KnownTimeGranularity["yearly"] = "yearly";
    /**
     * Monthly time granularity
     */
    KnownTimeGranularity["monthly"] = "monthly";
    /**
     * Weekly time granularity
     */
    KnownTimeGranularity["weekly"] = "weekly";
    /**
     * Daily time granularity
     */
    KnownTimeGranularity["daily"] = "daily";
    /**
     * Hourly time granularity
     */
    KnownTimeGranularity["hourly"] = "hourly";
    /**
     * Per minute time granularity
     */
    KnownTimeGranularity["perMinute"] = "minutely";
    /**
     * Per second time granularity
     */
    KnownTimeGranularity["perSecond"] = "secondly";
    /**
     * Per microsecond time granularity
     */
    KnownTimeGranularity["perMicrosecond"] = "microsecond";
    /**
     * None
     */
    KnownTimeGranularity["none"] = "none";
})(KnownTimeGranularity || (KnownTimeGranularity = {}));
